File: frontend/src/lib/feed.ts
import { api } from "./api";
import type { MediaType } from "./media";

export type ActivityAction = "favorited" | "status_changed" | "rated" | "commented";

export type Activity = {
  id: string;
  user: { id: string; username: string };
  media: {
    tmdb_id: number;
    media_type: MediaType;
    title: string;
    poster_url: string | null;
  };
  action: ActivityAction;
  detail: string | null;
  created_at: string;
};

export type FeedPage = {
  results: Activity[];
  page: number;
  has_more: boolean;
};

export async function getFeed(page: number): Promise<FeedPage> {
  const { data } = await api.get<FeedPage>("/feed", { params: { page, page_size: 20 } });
  return data;
}

File: backend/app/services/feed.py
"""Feed social: atividades recentes dos amigos aceitos do usuário."""
from sqlalchemy.orm import Session

from app.models.activity import Activity
from app.models.media import Media
from app.models.user import User
from app.services.friends import get_friend_ids


def list_feed(db: Session, user_id, page: int = 1, page_size: int = 20) -> dict:
    friend_ids = get_friend_ids(db, user_id)
    if not friend_ids:
        return {"results": [], "page": page, "has_more": False}

    offset = (page - 1) * page_size
    rows = (
        db.query(Activity, Media, User)
        .join(Media, Activity.media_id == Media.id)
        .join(User, Activity.user_id == User.id)
        .filter(Activity.user_id.in_(friend_ids))
        .order_by(Activity.created_at.desc())
        .offset(offset)
        .limit(page_size + 1)
        .all()
    )

    has_more = len(rows) > page_size
    rows = rows[:page_size]

    results = [
        {
            "id": str(activity.id),
            "user": {"id": str(user.id), "username": user.username},
            "media": {
                "tmdb_id": media.tmdb_id,
                "media_type": media.media_type,
                "title": media.title,
                "poster_url": media.poster_url,
            },
            "action": activity.action,
            "detail": activity.detail,
            "created_at": activity.created_at,
        }
        for activity, media, user in rows
    ]
    return {"results": results, "page": page, "has_more": has_more}
